pydoctor/templatewriter/__init__.py

```diff
@@ -5,6 +5,364 @@
+                    raise UnsupportedTemplateVersion(f"It appears that your custom template '{template.name}' is designed for a newer version of pydoctor."
+                                        "Rendering will most probably fail. Upgrade to latest version of pydoctor with 'pip install -U pydoctor'. ")
+        finally:
+            self._templates[template.name] = template
+
+
+    def add_templatedir(self, dir: Path) -> None:
+        """
+        Scan a directory and add all templates in the given directory to the lookup. 
+        """
+        for path in dir.iterdir():
+            template = Template.fromfile(path)
+            if template:
+                self.add_template(template)
+
+
+    def get_template(self, filename: str) -> Template:
+        """
+        Lookup a template based on its filename. 
+
+        Return the custom template if provided, else the default template.
+
+        @param filename: File name, (ie 'index.html')
+        @return: The Template object
+        @raises KeyError: If no template file is found with the given name
+        """
+        try:
+            t = self._templates[filename]
+        except KeyError as e:
+            raise KeyError(f"Cannot find template '{filename}' in template lookup: {self}. "
+                f"Valid filenames are: {list(self._templates)}") from e
+        return t
+
+class TemplateElement(Element, abc.ABC):
+    """
+    Renderable element based on a template file. 
+    """
+    def __init__(self, loader: ITemplateLoader ) -> None:
+        """
+        Init a new element. 
+
+        @raises TypeError: If C{loader} is not a L{ITemplateLoader} provider. 
+        """
+        try:
+            verify.verifyObject(ITemplateLoader, loader)
+        except Exception as e:
+            raise TypeError(f"Cannot create HTML element {self} because template loader "
+                            f"is not a ITemplateLoader provider: {type(loader)}") from e
+        super().__init__(loader)
+
+    @classmethod
+    def lookup_loader(cls, template_lookup: TemplateLookup) -> ITemplateLoader:
+        """
+        Lookup the template loader with the the C{TemplateLookup}. 
+
+        @raise TypeError: If the C{filename} property is not set. 
+        """
+        if cls.filename:
+            template = cls.template(template_lookup)
+            loader = template.loader
+            assert loader is not None
+            return loader
+        else:
+            raise TypeError(f"Cannot create loader because '{cls}.filename' property is not set.")
+
+    @classmethod
+    def template(cls, template_lookup: Optional[TemplateLookup] = None) -> Template:
+        # error: Argument 1 to "get_template" of "TemplateLookup" has incompatible type
+        # "Callable[[TemplateElement], str]"; expected "str"
+        # We can ignore mypy error because filename is an abstractproperty. 
+        if not template_lookup:
+            template_lookup = TemplateLookup()
```

Comment:
Could we use a single default `TemplateLookup` object instead of constructing a new one every time? The initialization of `TemplateLookup` looks relatively expensive.

Reply:
The optional value is only used in tests

Reply:
I could remove it all together

Reply:
The argument is not optional anymore